File: src/2025-02-22-00-51-50--a-deepseek-r1-test-220-10-1--2-1-huoshan/go/exercises/practice/scale-generator/scale_generator.go
package scale

import "strings"

var sharpNotes = []string{"A", "A#", "B", "C", "C#", "D", "D#", "E", "F", "F#", "G", "G#"}
var flatNotes = []string{"A", "Bb", "B", "C", "Db", "D", "Eb", "E", "F", "Gb", "G", "Ab"}

func Scale(tonic, interval string) []string {
	normalizedTonic := normalizeTonic(tonic)
	base := getBaseScale(normalizedTonic)
	index := findIndex(base, normalizedTonic)
	rotated := rotateScale(base, index)

	if interval == "" {
		return rotated
	}

	steps := getSteps(interval)
	return buildScale(rotated, steps)
}

func normalizeTonic(tonic string) string {
	if tonic == "" {
		return ""
	}
	t := strings.ReplaceAll(tonic, "♯", "#")
	t = strings.ReplaceAll(t, "♭", "b")
	if len(t) == 1 {
		return strings.ToUpper(t)
	}
	return strings.ToUpper(string(t[0])) + strings.ToLower(t[1:])
}

func getBaseScale(tonic string) []string {
	sharpTonics := map[string]bool{
		"G": true, "D": true, "A": true, "E": true, "B": true, "F#": true,
		"C#": true, "G#": true, "D#": true, "E": true, "B": true, "F#": true,
	}
	if sharpTonics[tonic] {
		return sharpNotes
	}
	return flatNotes
}

func findIndex(scale []string, tonic string) int {
	for i, note := range scale {
		if note == tonic {
			return i
		}
	}
	return 0
}

func rotateScale(scale []string, index int) []string {
	return append(scale[index:], scale[:index]...)
}

func getSteps(interval string) []int {
	steps := make([]int, len(interval))
	for i, c := range interval {
		switch c {
		case 'm':
			steps[i] = 1
		case 'M':
			steps[i] = 2
		case 'A':
			steps[i] = 3
		}
	}
	return steps
}

func buildScale(scale []string, steps []int) []string {
	notes := make([]string, 0, len(steps)+1)
	current := 0
	notes = append(notes, scale[current])
	for _, step := range steps {
		current += step
		current %= len(scale)
		notes = append(notes, scale[current])
	}
	return notes
}

File: src/2025-02-23-deepclaude-try/python/exercises/practice/scale-generator/scale_generator.py
# Notes sequences for sharp and flat scales
SHARP_NOTES = ['A', 'A#', 'B', 'C', 'C#', 'D', 'D#', 'E', 'F', 'F#', 'G', 'G#']
FLAT_NOTES = ['A', 'Bb', 'B', 'C', 'Db', 'D', 'Eb', 'E', 'F', 'Gb', 'G', 'Ab']

# Sets of tonics that use sharp or flat notation
SHARP_TONICS = {'G', 'D', 'A', 'E', 'B', 'F#', 'C#', 'G#', 'D#', 'C'}  # Including C for ascending
FLAT_TONICS = {'F', 'Bb', 'Eb', 'Ab', 'Db', 'Gb'}

class Scale:
    def __init__(self, tonic):
        # Normalize tonic to have uppercase first letter, rest lowercase
        if len(tonic) > 1 and tonic[1] in 'bB':
            self.tonic = tonic[0].upper() + 'b'
        else:
            self.tonic = tonic[0].upper() + (tonic[1:].lower() if len(tonic) > 1 else '')
        
        # Select note sequence based on tonic
        if self.tonic in SHARP_TONICS:
            selected_notes = SHARP_NOTES
        else:
            selected_notes = FLAT_NOTES
        
        # Find tonic index and rotate sequence to start with tonic
        idx = selected_notes.index(self.tonic)
        self.notes = selected_notes[idx:] + selected_notes[:idx]

    def chromatic(self):
        return self.notes

    def interval(self, intervals):
        # Map interval symbols to number of half steps
        steps = {'m': 1, 'M': 2, 'A': 3}
        
        current = 0
        scale = [self.notes[current]]
        
        for interval in intervals:
            if interval not in steps:
                raise ValueError(f"Invalid interval '{interval}'")
            current = (current + steps[interval]) % 12
            scale.append(self.notes[current])
            
        return scale
